### DISCUSSION TITLE
Navigation with pop and push animations

### DISCUSSION THREAD

**Discussion by alinless** [pointfreeco/swift-composable-architecture#2234] [Q&A] [answered]
Hi,
I want to implement a simple navigation feature in my app. Specifically, I want to pop the second view and push the third view when the user taps a button on the second view.
Is it possible to achieve this using tree-based navigation? Or is it only possible with a stack-based approach?

FirstFeature action
```swift
         case .destination(.presented(.second(.delegate(.thirdRequested)))):
            state.destination = .third(.init())
            return .none
``` 

FirstView
```swift
      NavigationStack {
         ZStack {
            Color.green

            WithViewStore(self.store, observe: { $0 }) { viewStore in
               VStack {
                  Text("FirstView")
                  
                  Button.init {
                     viewStore.send(.showSecond)
                  } label: {
                     Text("To second view")
                  }
                  
               }
            }
         }
         .navigationDestination(
            store: self.store.scope(state: \.$destination, action: { .destination($0) }),
            state: /FirstFeature.Destination.State.second,
            action: FirstFeature.Destination.Action.second,
            destination: SecondView.init(store:)
         )
         .navigationDestination(
            store: self.store.scope(state: \.$destination, action: { .destination($0) }),
            state: /FirstFeature.Destination.State.third,
            action: FirstFeature.Destination.Action.third,
            destination: ThirdView.init(store:)
         )
      }
``` 

https://github.com/pointfreeco/swift-composable-architecture/assets/94690815/e45a001d-43a3-48cc-8c9c-c133120b276a

**Comment by mbrandonw** [comment]
Hi @alinless, based on your little code snippet I think the app is actually behaving correctly. The problem is that when you mutate the `destination` to point from `second` to `third`, you are technically popping the second view off the navigation since that state is being cleared out. In order to preserve the second screen while also pushing the third screen, you would need to move the third screen navigation to the second screen's domain.

Or, if you really do want to coalesce all destinations at the root of the navigation view, then unfortunately you do need to use `NavigationStack`.

**Reply by alinless** [reply]
Thank you for the answer!
So, is it possible with a StackState to pop a second view and after push a third view with animations?

**Reply by mbrandonw** [reply]
Oh I'm sorry, I didn't read carefully enough. I now see that you specifically want to pop the second and push the third.

Yes, unfortunately `navigationDestination(isPresented:)` in vanilla SwiftUI cannot do that properly, and that's the API we use under the hood of `navigationDestination(store:)`. Really it's a bug in SwiftUI and should be reported.

If you want animation for a simultaneous push+pop I think you will have to use a navigation stack.

**Reply by alinless** [reply]
I have tried this method, but the result was still the same. There was no animation. How can I make it work?


```swift

         case .showSecond:
            state.path.append(.second())
            return .none

         case let .path(.element(id: _, action: .second(.delegate(action)))):
            switch action {
            case .thirdRequested:
               state.path.popLast()
               state.path.append(.third())
            }
            return .none
``` 

```swift
      NavigationStackStore(
         self.store.scope(state: \.path, action: { .path($0) })
      ) {
            ZStack {
               Color.green
               WithViewStore(self.store, observe: { $0 }) { viewStore in
                  VStack {
                     Text("FirstView")
                     
                     Button.init {
                        viewStore.send(.showSecond)
                     } label: {
                        Text("To second view")
                     }
                     
                  }
               }
            }
         } destination: { state in
            switch state {
            case .second:
               CaseLet(
                  state: /FirstFeature.Path.State.second,
                  action: FirstFeature.Path.Action.second,
                  then: SecondView.init(store:)
               )
            case .third:
               CaseLet(
                  state: /FirstFeature.Path.State.third,
                  action: FirstFeature.Path.Action.third,
                  then: ThirdView.init(store:)
               )
            }
         }
```

**Reply by mbrandonw** [reply]
Can you share your project?

**Reply by mbrandonw** [reply, marked as answer]
Hi @alinless, I just checked in vanilla SwiftUI and apparently that is the behavior there too. I thought for some reason it would animate but that is not the case.

Here's a sample for you to test:

```swift
import SwiftUI

class Model: ObservableObject {
  @Published var numbers: [Int] = []

  func push() {
    self.numbers.append(.random(in: 1...100))
  }
  func popPush() {
    self.numbers.removeLast()
    self.numbers.append(.random(in: 1...100))
  }
}

struct ContentView: View {
  @ObservedObject var model = Model()
  var body: some View {
    NavigationStack(path: self.$model.numbers) {
      Button("Push") {
        self.model.push()
      }
      .navigationDestination(for: Int.self) { int in
        NumberView(number: int, model: self.model)
      }
    }
  }
}

struct NumberView: View {
  let number: Int
  let model: Model
  var body: some View {
    VStack {
      Text("\(self.number)")
      Button("Pop/push") {
        self.model.popPush()
      }
    }
  }
}

#Preview {
  ContentView()
}
```

**Reply by alinless** [reply]
Thanks

https://github.com/alinless/TestTCANavigation
Here is an example with a TCA StackState and a similar problem. In this case, the solution seems to be using a delay between pop and push (appending to a stack). It would be great if there are other options to handle it

**Reply by mbrandonw** [reply]
The library doesn't try to fix deficiencies in SwiftUI, and so because this is how vanilla SwiftUI works I don't think there is anything we can do.

**Reply by alinless** [reply]
Got it. Anyway, thank you very much!
